google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java

```diff
@@ -686,6 +710,29 @@ public EnhancedBigtableStub(EnhancedBigtableStubSettings settings, ClientContext
   }
   // </editor-fold>
 
+  /**
+   * Adds a HeaderTracer instance and current span name to CallOptions so we could surface metrics
+   * in the header with {@link
+   * com.google.cloud.bigtable.data.v2.stub.metrics.ClientHeaderInterceptor}.
+   */
+  private ApiCallContext getContextWithTracer(SpanName spanName) {
+    ApiCallContext apiCallContext = clientContext.getDefaultCallContext();
+    if (!(apiCallContext instanceof GrpcCallContext)) {
+      LOGGER.warning(
+          "Failed to add tracer in call context. Expected GrpcCallContext but had "
+              + apiCallContext.getClass());
+      return apiCallContext;
+    }
+    GrpcCallContext grpcCallContext = (GrpcCallContext) apiCallContext;
+    CallOptions options = grpcCallContext.getCallOptions();
+    options =
+        options
+            .withOption(HeaderTracer.HEADER_TRACER_CONTEXT_KEY, settings.getHeaderTracer())
+            .withOption(HeaderTracer.SPAN_NAME_CONTEXT_KEY, spanName.toString());
```

Comment:
I would try to wrap both values in a single object to minimize the overhead of adding CallOptions entries which is backed by an array